### PR TITLE
Add Razor Pages 2FA support

### PR DESCRIPTION
**Pull request by danroth27**
@DamianEdwards @blowdart

### CODE REVIEW

File: template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/LoginWith2fa.cshtml.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+using Microsoft.AspNetCore.Identity;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.RazorPages;
+using Microsoft.AspNetCore.Mvc.ViewFeatures;
+using Microsoft.Extensions.Logging;
+using Company.WebApplication1.Data;
+
+namespace Company.WebApplication1.Pages.Account
+{
+    public class LoginWith2faModel : PageModel
```
Comment (HaoK) [inline template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/LoginWith2fa.cshtml.cs]: With?  Why not just Login2faModel or 2faLoginModel?

File: template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/LoginWith2fa.cshtml.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                return LocalRedirect(Url.GetLocalUrl(returnUrl));
+            }
+            if (result.IsLockedOut)
+            {
+                _logger.LogWarning("{UserName} account locked out.", user.UserName);
+                return RedirectToPage("./Lockout");
+            }
+            else
+            {
+                _logger.LogWarning("Invalid authenticatoe code entered for {UserName}.", user.UserName);
```
Comment (HaoK) [inline template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/LoginWith2fa.cshtml.cs]: typo

File: template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/LoginWithRecoveryCode.cshtml.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.RazorPages;
+using Microsoft.Extensions.Logging;
+using Company.WebApplication1.Data;
+
+namespace Company.WebApplication1.Pages.Account
+{
+    public class LoginWithRecoveryCodeModel : PageModel
```
Comment (HaoK) [inline template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/LoginWithRecoveryCode.cshtml.cs]: LoginRecoveryCodeModel/RecoveryCodeLoginModel?

File: template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/Disable2fa.cshtml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+<h2>@ViewData["Title"].</h2>
+
+<div class="alert alert-warning" role="alert">
+    <p>
+        <span class="glyphicon glyphicon-warning-sign"></span>
+        <strong>This action only disables 2FA.</strong>
+    </p>
+    <p>
+        Disabling 2FA does not change the keys used in authenticator apps. If you wish to change the key
```
Comment (HaoK) [inline template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/Disable2fa.cshtml]: Should we be consistent with casing? 2fa / 2FA

File: template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/LoginWith2fa.cshtml.cs (hunk 1 above)
Comment (blowdart) [inline template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/LoginWith2fa.cshtml.cs]: Class names beginning with numbers are the work of the devil.

It's in keeping with the page name, which seems to be the standard here.

File: template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/LoginWithRecoveryCode.cshtml.cs (hunk 3 above)
Comment (blowdart) [inline template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/LoginWithRecoveryCode.cshtml.cs]: Again, in keeping with the page name

File: template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/Disable2fa.cshtml (hunk 4 above)
Comment (blowdart) [inline template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/Disable2fa.cshtml]: Yea, lower is better.

File: template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/Enable2fa.cshtml.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            var user = await _userManager.GetUserAsync(HttpContext.User);
+            if (user == null)
+            {
+                return RedirectToPage("/Error");
+            }
+
+            if (await _userManager.GetTwoFactorEnabledAsync(user))
+            {
+                return RedirectToPage("/Error");
```
Comment (HaoK) [inline template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/Enable2fa.cshtml.cs]: Should we actually tell the user what the error is?

File: template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/Enable2fa.cshtml.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            }
+
+            if (!(await _userManager.GetValidTwoFactorProvidersAsync(user)).Any())
+            {
+                return RedirectToPage("/Error");
```
Comment (HaoK) [inline template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/Enable2fa.cshtml.cs]: Should we actually tell the user what the error is?

File: template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/EnableAuthenticator.cshtml.cs [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+            }
+
+            return result.ToString().ToLowerInvariant();
+        }
+
+        private string GenerateQrCodeUri(string email, string unformattedKey)
+        {
+            return string.Format(
+                AuthenicatorUriFormat,
+                "Company.WebApplication1",
```
Comment (HaoK) [inline template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/EnableAuthenticator.cshtml.cs]: Can we calculate this off of a type/namespace/nameof so its not a random hardcoded string?

File: template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/LoginWith2fa.cshtml.cs (hunk 1 above)
Comment (HaoK) [inline template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/LoginWith2fa.cshtml.cs]: Yeah the suggestion is that you rename both to get rid of With :)

File: template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/EnableAuthenticator.cshtml.cs (hunk 7 above)
Comment (blowdart) [inline template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/EnableAuthenticator.cshtml.cs]: It should come from the name you give when you create the template, like the copyright in the layout. So we're consistent here with everything else.

File: template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/Enable2fa.cshtml.cs (hunk 5 above)
Comment (blowdart) [inline template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/Enable2fa.cshtml.cs]: The only way to get here in the already enabled state is to type the URL, so, meh, not really.

File: template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/Enable2fa.cshtml.cs (hunk 6 above)
Comment (blowdart) [inline template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/Enable2fa.cshtml.cs]: The only way to get here in the already enabled state is to type the URL, so, meh, not really.

File: template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/Disable2fa.cshtml (hunk 4 above)
Comment (danroth27) [inline template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/Disable2fa.cshtml]: In prose, 2FA is an acronym and the consensus on the web appears to be that it should be all caps.

File: template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/EnableAuthenticator.cshtml.cs (hunk 7 above)
Comment (danroth27) [inline template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/EnableAuthenticator.cshtml.cs]: "Company.WebApplication1" gets replaced by the template engine with the project name.

File: template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/Disable2fa.cshtml (hunk 4 above)
Comment (HaoK) [inline template_feed/Microsoft.DotNet.Web.ProjectTemplates.2.0/content/RazorPagesWeb-CSharp/Pages/Account/Manage/Disable2fa.cshtml]: Yeah personally lower looked weird to me, all caps acronyms seems better

**Comment by danroth27** [comment]
@mlorbetske Should be ready to go

**Pull request merged**